### ISSUE TITLE
The missing documentation for using the mosquitto "snap" package on Ubuntu

### ISSUE THREAD

**Issue by bootrino** [eclipse/mosquitto#931]
OK it cost me many hours of pain.  Hopefully this will save you some time, if you are using the latest version of mosquitto on Ubuntu, which is packaged as a "snap" and therefore the configuration does NOT happen in /etc.

Unfortunately there is some logic in /snap/mosquitto/51/launcher.sh that makes it very unclear exactly where to configure mosquitto and you'll have to read that script pretty much any time you need to configure mosquitto - I think this configuration logic should have been completely avoided and would have been better if the config just went into one single well defined place.  Anyhow.

Here is the missing documentation:

####Using Mosquitto as a "snap" file on Ubuntu

#to work out the location of the mosquitto.conf file being used, reverse engineer the source of this file:
cat /snap/mosquitto/51/launcher.sh

# after some extensive fiddling I found putting the configuration file here seems to work:
/var/snap/mosquitto/common/mosquitto.conf
note that you might need to delete all others files from this directory to ensure the launcher.sh is not picking them up and changing the configuration because of their presence.

# general tip: to list the systemd mosquitto service unit files
systemctl list-unit-files | grep mosquitto

# general tip: to stop the mosquitto service:
sudo systemctl stop snap.mosquitto.mosquitto.service

# general tip: to start the mosquitto service
sudo systemctl start snap.mosquitto.mosquitto.service

# general tip: to find out the current status of the mosquitto service
sudo systemctl status snap.mosquitto.mosquitto.service

# general tip: to see the mosquitto service log messages:
sudo journalctl -f -u snap.mosquitto.mosquitto.service

# apparently you will need to run mosquitto as root

# if you do not set the user in the conf file then you will get the following errors in the systemd log:

Aug 23 14:43:44 ip-10-0-0-30 mosquitto.mosquitto[10690]: 1535060624: Error: Invalid user 'mosquitto'.

# if you set the user ubuntu you will get this message:
Aug 23 14:45:12 ip-10-0-0-30 mosquitto.mosquitto[10968]: 1535060712: Error setting groups whilst dropping privileges: Operation not permitted.

# if you set the user to root, it will work, but you will get this message:
Aug 23 14:46:32 ip-10-0-0-30 mosquitto.mosquitto[11165]: 1535060792: Warning: Mosquitto should not be run as root/administrator.

# mosquitto will not start if your config file has include_dir /etc/mosquitto/conf.d because it cannot find the directory

****** to do a test message publish, on the same machine as the server:
mosquitto_sub -h localhost -t test
# and on the same macine in a different console:
mosquitto_pub -h localhost -t test -m atestmessage

**Comment by bootrino** [comment]
Is /etc no longerused with snap files?  It's pretty annoying if snaps change /etc  it's a reasonable assumption that config is in /etc for pretty much anything.

**Comment by ralight** [comment]
That's right, a normal snap package only has access to specific paths, and that doesn't include anything in /etc.

**Comment by bootrino** [comment]
Feel free to move this to the documentation.

**Comment by Sequenzial** [comment]
Does anyone know how to clean the db? On normal installation I see a mosquitto.db that can be deleted while the service is down. But on Snapinstall there is no mosquitto.db  (or I cannot find it)

**Comment by Sequenzial** [comment]
Makes sence! 
Thank you for this Information!
